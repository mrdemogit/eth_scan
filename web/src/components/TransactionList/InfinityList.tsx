import React, { Component } from 'react';
import styled from 'styled-components/macro';
import { TransactionType } from 'transactionTypes';
import InfiniteScroll from 'react-infinite-scroller';
import { OFFSET } from '../../Datasource';

const ListStyled = styled.div`
  background: #353535;
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: column;
`;

interface Props {
  transactions: TransactionType[];
  children: (values: TransactionType[]) => React.ReactNode;
}

interface State {
  renderedTransactions: TransactionType[];
}

class InfinityList extends Component<Props, State> {
  state = {
    renderedTransactions: this.props.transactions,
  };

  loadMore = (page: number) => {
    console.log('fired', page);
    const { transactions } = this.props;
    const start = page * OFFSET;
    const end = start + OFFSET;
    this.setState(({ renderedTransactions }) => ({
      renderedTransactions: [
        ...renderedTransactions,
        ...transactions.slice(start, end),
      ],
    }));
  };

  render() {
    const { renderedTransactions } = this.state;
    const { children, transactions } = this.props;
    return (
      <ListStyled>
        <InfiniteScroll
          pageStart={0}
          loadMore={this.loadMore}
          hasMore={renderedTransactions.length !== transactions.length}
          loader={<div key={0}>Loading ...</div>}
        >
          {children(renderedTransactions)}
        </InfiniteScroll>
      </ListStyled>
    );
  }
}

export default InfinityList;
