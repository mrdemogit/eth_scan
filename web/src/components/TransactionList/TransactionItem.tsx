import React, { memo } from 'react';
import styled from 'styled-components/macro';
import { TransactionType } from 'transactionTypes';
import moment from 'moment';
import { formatEthValue } from '@utils/index';

const formatTime = (timeStamp: number) => {
  const txTime = moment(timeStamp * 1000);
  const weekLimit = moment().add(-1, 'week');
  const currentYear = moment().year();
  const txYear = txTime.year();
  if (currentYear !== txYear) {
    return txTime.format('MMM YYYY');
  }
  return txTime.isBefore(weekLimit)
    ? txTime.format('MMM DD')
    : txTime.fromNow();
};

const ItemStyled = styled.div`
  display: flex;
  padding: 20px 0;
  margin: 0 20px;
  max-width: 480px;
  border-bottom: 1px solid #5f5f5f;
  width: 100%;
  text-align: left;
`;

const CurrencyStyled = styled.div`
  border-radius: 3rem;
  background: linear-gradient(#baa8a8, #958686);
  padding: 13px 0;
  font-size: 17px;
  min-width: 45px;
  height: 19px;
  margin: 0.3rem 1rem;
  text-align: center;
`;

const ContentStyled = styled.div`
  width: calc(100% - 90px);
`;

const ValueStyled = styled.div`
  font-size: 20px;
`;

const AddressStyled = styled.div`
  text-transform: uppercase;
  margin: 0.2rem 0;
  font-size: 12px;
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
  padding-right: 10px;
`;

const ValueAndTime = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
`;
const TimeStyled = styled.div`
  color: #a5a5a5;
  font-size: 14px;
`;

const Arrow = styled.span`
  transform: rotate(45deg);
  content: ' ';
  display: block;
  width: 8px;
  height: 8px;
  float: right;
  border: 2px solid #afafaf;
  border-bottom: none;
  border-left: none;
  margin-top: 3px;
  margin-left: 8px;
`;

const TransactionItem: React.FC<TransactionType> = ({
  value,
  hash,
  timeStamp,
  from,
  to,
}) => (
  <ItemStyled>
    <CurrencyStyled>ETH</CurrencyStyled>
    <ContentStyled>
      <ValueAndTime>
        <ValueStyled>{formatEthValue(value)} ETH</ValueStyled>
        <TimeStyled>
          {formatTime(parseInt(timeStamp, 10))} <Arrow />
        </TimeStyled>
      </ValueAndTime>
      {hash === from ? (
        <React.Fragment>
          <AddressStyled>Sent to</AddressStyled>
          <AddressStyled>{to}</AddressStyled>
        </React.Fragment>
      ) : (
        <React.Fragment>
          <AddressStyled>Received from</AddressStyled>
          <AddressStyled>{from}</AddressStyled>
        </React.Fragment>
      )}
    </ContentStyled>
  </ItemStyled>
);

export default memo(TransactionItem);
