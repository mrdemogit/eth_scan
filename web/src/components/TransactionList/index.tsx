import React from 'react';
import styled from 'styled-components/macro';
import { TransactionType } from 'transactionTypes';

const ListStyled = styled.div``;

const ErrorStyled = styled.div`
  color: red;
  padding: 1rem;
`;

const LoadingStyled = styled.div`
  padding: 1rem;
`;

interface Props {
  error?: string;
  fetching: boolean;
  transactions: TransactionType[];
}

const TransactionList: React.FC<Props> = ({ error, transactions }) => {
  if (error) {
    return <ErrorStyled>{error}</ErrorStyled>;
  }
  if (!transactions) {
    return <LoadingStyled>Loading...</LoadingStyled>;
  }
  return <ListStyled>gagea</ListStyled>;
};

export default TransactionList;
