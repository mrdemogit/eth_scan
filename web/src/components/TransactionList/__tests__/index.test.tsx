import React from 'react';
import { render } from 'react-testing-library';
import TransactionList from '../index';

describe('TransactionList', () => {
  jest.mock('moment', () => () => ({
    format: () => '2018–01–30T12:34:56+00:00',
  }));

  const formatTime = () => 'time';
  it('Render empty TransactionList', () => {
    const { container } = render(
      <TransactionList error={null} transactions={[]} isFetching={false} />,
    );
    expect(container).toMatchSnapshot();
  });

  it('Render error TransactionList', () => {
    const { container } = render(
      <TransactionList
        error={'Error!!!!'}
        transactions={[]}
        isFetching={false}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('Render Loading TransactionList', () => {
    const { container } = render(
      <TransactionList error={null} transactions={[]} isFetching={true} />,
    );
    expect(container).toMatchSnapshot();
  });

  it('Render Loading TransactionList', () => {
    const { container } = render(
      <TransactionList
        error={null}
        transactions={[
          {
            value: '3234',
            hash: 'grsg42t4$$24',
            timeStamp: 1550415453,
            from: 'fromaddress',
            to: 'toAddress',
          },
        ]}
        isFetching={false}
      />,
    );
    expect(container).toMatchSnapshot();
  });
});
