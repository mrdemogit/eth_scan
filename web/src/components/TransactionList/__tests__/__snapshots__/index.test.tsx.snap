// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionList Render Loading TransactionList 1`] = `
<div>
  <div
    class="TransactionList__LoadingStyled-sc-1k4q6wn-1 eckYlr"
  >
    Loading...
  </div>
</div>
`;

exports[`TransactionList Render Loading TransactionList 2`] = `
<div>
  <div
    class="InfinityList__ListStyled-sc-7l0oum-0 cTFxTz"
  >
    <div>
      <div
        class="TransactionItem__ItemStyled-sc-1199a3v-0 fpAPjr"
      >
        <div
          class="TransactionItem__CurrencyStyled-sc-1199a3v-1 dAzMgI"
        >
          ETH
        </div>
        <div
          class="TransactionItem__ContentStyled-sc-1199a3v-2 lhVVNx"
        >
          <div
            class="TransactionItem__ValueAndTime-sc-1199a3v-5 eenYPy"
          >
            <div
              class="TransactionItem__ValueStyled-sc-1199a3v-3 bJYuN"
            >
              3.234e-15
               ETH
            </div>
            <div
              class="TransactionItem__TimeStyled-sc-1199a3v-6 fGYyVv"
            >
              2018–01–30T12:34:56+00:00
               
              <span
                class="TransactionItem__Arrow-sc-1199a3v-7 kuRVia"
              />
            </div>
          </div>
          <div
            class="TransactionItem__AddressStyled-sc-1199a3v-4 hLdONn"
          >
            RECEIVED FROM
          </div>
          <div
            class="TransactionItem__AddressStyled-sc-1199a3v-4 hLdONn"
          >
            fromaddress
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`TransactionList Render empty TransactionList 1`] = `
<div>
  <div
    class="InfinityList__ListStyled-sc-7l0oum-0 cTFxTz"
  >
    <div />
  </div>
</div>
`;

exports[`TransactionList Render error TransactionList 1`] = `
<div>
  <div
    class="TransactionList__ErrorStyled-sc-1k4q6wn-0 bDzcTg"
  >
    Error!!!!
  </div>
</div>
`;
